File: DepthFirstSearch.java
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class DepthFirstSearch {
    private Set<String> marked;
    private Map<String, String> edgeTo;
    private String s;

    public DepthFirstSearch(Graph g, String s) {
        this.s = s;
        marked = new HashSet<>();
        edgeTo = new HashMap<>();
        dfs(g, s);
    }

    public boolean hasPathTo(String v) {
        return marked.contains(v);
    }

    public Iterable<String> pathTo(String v) {
        List<String> path = new LinkedList<>();
        if (hasPathTo(v)) {
            while (!v.equals(s)) {
                path.add(0, v);
                v = edgeTo.get(v);
            }
            path.add(0, s);
        }
        return path;
    }

    private void dfs(Graph g, String v) {
        marked.add(v);
        for (String w : g.getAdj(v)) {
            if (!marked.contains(w)) {
                edgeTo.put(w, v);
                dfs(g, w);
            }
        }
    }
}

File: depthfirstsearch.cpp
#include "depthfirstsearch.h"
#include <iostream>
#include <fstream>
#include <sstream>
#include <set>

using namespace std;

DepthFirstSearch::DepthFirstSearch(Graph &g, string s)
{
	this->s = s;
	dfs(g, s);
}

void DepthFirstSearch::dfs(Graph &g, string v)
{
	// cout << "em " << v << endl;
	marked.insert(v); // marca v como visitado
	for (auto const &w : g.getAdj(v))
	{
		if (marked.find(w) == marked.end())
		{
			edgeTo[w] = v;
			dfs(g, w);
		}
	}
}

bool DepthFirstSearch::hasPathTo(string s)
{
	return marked.find(s) != marked.end();
}

vector<string> DepthFirstSearch::pathTo(string v)
{
	vector<string> path;
	while (v != s)
	{
		path.insert(path.begin(), v);
		v = edgeTo[v];
	}
	path.insert(path.begin(), s);
	return path;
}
